README.md
## 正则表达式插件(vscode)

方便查找，具体的 Vscode 插件学习详见[小茗同学](https://www.cnblogs.com/liuxianan/p/vscode-plugin-overview.html)

## 用法

光标置于需要添加的地方 `Ctrl + shift + p` 出现输入框，输入关键字(例如：车牌号) 或者 `wqjiao`,点击需要的正则。

![图片加载中...](https://raw.githubusercontent.com/wqjiao/regular-set/master/assets/use-reg.gif)

## 表达式

* 手机号码(严谨)

```js
/^1((3[\d])|(4[5,6,7,9])|(5[0-3,5-9])|(6[5-7])|(7[0-8])|(8[\d])|(9[1,8,9]))\d{8}$/
```

* 手机号码(宽松)

```js
/^1[3-9]\d{9}$/
```

* 国内座机电话

```js
/^\d{3}-\d{8}|\d{4}-\d{7}$/
```

* 座机电话(包括港澳台)

```js
/^(\d{3}-\d{8})|(\d{4}-\d{7})|(852|853|886-\d{7,8})$/
```

* 电子邮箱

```js
/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
```

* 身份证号码

```js
/^\d{6}(18|19|([23]\d))\d{2}((0[1-9])|(10|11|12))(([0-2][1-9])|10|20|30|31)\d{3}[0-9Xx]$/
```

* 护照(包含香港、澳门)

```js
/(^[EeKkGgDdSsPpHh]\d{8}$)|(^(([Ee][a-fA-F])|([DdSsPp][Ee])|([Kk][Jj])|([Mm][Aa])|(1[45]))\d{7}$)/
```

* 车牌号(新能源+非新能源)

```js
/^([京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}(([0-9]{5}[DF])|([DF]([A-HJ-NP-Z0-9])[0-9]{4})))|([京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-Z]{1}[A-HJ-NP-Z0-9]{4}[A-HJ-NP-Z0-9 挂学警港澳]{1})$/
```

* 新能源车牌号

```js
/[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}(([0-9]{5}[DF])|([DF][A-HJ-NP-Z0-9][0-9]{4}))$/
```

* 非新能源车牌号

```js
/^[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}[A-Z0-9]{4}[A-Z0-9挂学警港澳]{1}$/
```

* 车架号

```js
/^[A-Z0-9]{17}$/
```

* 发动机号

```js
/^[A-Z0-9]{6,17}$/
```

* 中国邮政编码

```js
/^(0[1-7]|1[0-356]|2[0-7]|3[0-6]|4[0-7]|5[1-7]|6[1-7]|7[0-5]|8[013-6])\d{4}$/
```

* 允许字母数字下划线组合

```js
/^[a-zA-Z0-9_]{6,16}$/
```

* 汉字

```js
/^[\u4E00-\u9FA5]+$/
```

* 英文

```js
/^[a-zA-Z]+$/
```

* 数字

```js
/^\d{1,}$/
```

* 小数

```js
/^\d+\.\d+$/
```

* 不能包含汉字

```js
/^[^\u4E00-\u9FA5]*$/
```


CHANGELOG.md
# 更改日志

便于统一项目的正则集合，不保证所有地方公用，如有错误，提交至：https://github.com/wqjiao/regular-set/issues

* 0.1.0 (2020-02-11)

    - 支持 Vscode 编辑器内 `Ctrl + Shift + p` 搜索：`wqjiao.`





src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
const REGULARS = require('../assets/REGULARS.js');

// 激活扩展程序时执行
function activate(context: vscode.ExtensionContext) {
	
	console.log('Congratulations, your extension "regular-set" is now active!');

	// 多条规则组合，遍历生成多条规则
	REGULARS.forEach(({ title, rule }: { title: string, rule: RegExp, example: string }, index: string) => {
		let disposable = vscode.commands.registerCommand(`extension.rule${index}`, () => {
			const editor = vscode.window.activeTextEditor;

			if (editor) {
				const { selections } = editor;

				editor.edit(editBuilder => {
					selections.forEach(selection => {
						const { start, end } = selection;
						const range = new vscode.Range(start, end);
						editBuilder.replace(range, String(rule));
					});
				});
				// 操作成功时，消息窗口提示
				vscode.window.showInformationMessage(`已插入正则: ${title}`);
			} else {
				// 操作错误时，消息窗口提示
				vscode.window.showWarningMessage('regular-set: 只有在编辑文本的时候才可以使用!');
			}
		});

		context.subscriptions.push(disposable);
	});
}

// 停止扩展程序时执行
function deactivate() {
	vscode.window.showWarningMessage('regular-set: 已关闭!');
}

module.exports = {
	activate,
	deactivate
}


assets/REGULARS.js
module.exports = [
    {
        title: '手机号码(严谨)',
        rule: /^1((3[\d])|(4[5,6,7,9])|(5[0-3,5-9])|(6[5-7])|(7[0-8])|(8[\d])|(9[1,8,9]))\d{8}$/,
        examples: ['18118360369'],
    },
    {
        title: '手机号码(宽松)',
        rule: /^1[3-9]\d{9}$/,
        examples: ['18118360369'],
    },
    {
        title: '国内座机电话',
        rule: /^\d{3}-\d{8}|\d{4}-\d{7}$/,
        examples: ['025-83192650'],
    },
    {
        title: '座机电话(包括港澳台)',
        rule: /^(\d{3}-\d{8})|(\d{4}-\d{7})|(852|853|886-\d{7,8})$/,
        examples: ['025-83192650'],
    },
    {
        title: '电子邮箱',
        rule: /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/,
        examples: ['888888888.@qq.com'],
    },
    {
        title: '身份证号码',
        rule: /^\d{6}(18|19|([23]\d))\d{2}((0[1-9])|(10|11|12))(([0-2][1-9])|10|20|30|31)\d{3}[0-9Xx]$/,
        examples: ['500222198105267734'],
    },
    {
        title: '护照(包含香港、澳门)',
        rule: /(^[EeKkGgDdSsPpHh]\d{8}$)|(^(([Ee][a-fA-F])|([DdSsPp][Ee])|([Kk][Jj])|([Mm][Aa])|(1[45]))\d{7}$)/,
        examples: [],
    },
    {
        title: '车牌号(新能源+非新能源)',
        rule: /^([京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}(([0-9]{5}[DF])|([DF]([A-HJ-NP-Z0-9])[0-9]{4})))|([京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-Z]{1}[A-HJ-NP-Z0-9]{4}[A-HJ-NP-Z0-9 挂学警港澳]{1})$/,
        examples: ['苏A1277J'],
    },
    {
        title: '新能源车牌号',
        rule: /[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}(([0-9]{5}[DF])|([DF][A-HJ-NP-Z0-9][0-9]{4}))$/,
        examples: ['京AD92035', '甘G23459F'],
    },
    {
        title: '非新能源车牌号',
        rule: /^[京津沪渝冀豫云辽黑湘皖鲁新苏浙赣鄂桂甘晋蒙陕吉闽贵粤青藏川宁琼使领 A-Z]{1}[A-HJ-NP-Z]{1}[A-Z0-9]{4}[A-Z0-9挂学警港澳]{1}$/,
        examples: ['京A00599', '黑D23908']
    },
    {
        title: '车架号',
        rule: /^[A-Z0-9]{17}$/, // 1：生产国家；2：制造商；3：汽车类型；9：校验位；10：出厂日期；11组装地；12-17 生产序号
        examples: ['LGXCG6CF3D0027519'],
    },
    {
        title: '发动机号',
        rule: /^[A-Z0-9]{6,17}$/,
        examples: ['246764K'],
    },
    {
        title: '中国邮政编码',
        rule: /^(0[1-7]|1[0-356]|2[0-7]|3[0-6]|4[0-7]|5[1-7]|6[1-7]|7[0-5]|8[013-6])\d{4}$/,
        examples: ['734500', '100101']
    },
    {
        title: '允许字母数字下划线组合',
        rule: /^[a-zA-Z0-9_]{6,16}$/,
        examples: ['wqjiao_3690'],
    },
    {
        title: '汉字',
        rule: /^[\u4E00-\u9FA5]+$/,
        examples: ['我们'],
    },
    {
        title: '英文',
        rule: /^[a-zA-Z]+$/,
        examples: ['wqjiao'],
    },
    {
        title: '数字',
        rule: /^\d{1,}$/,
        examples: ['3690'],
    },
    {
        title: '小数',
        rule: /^\d+\.\d+$/,
        examples: ['0.0', '0.09']
    },
    {
        title: '不能包含汉字',
        rule: /^[^\u4E00-\u9FA5]*$/,
        examples: ['wqjiao'],
    },
];
